### DISCUSSION TITLE
Missing parameters from model fits when using beta family

### DISCUSSION THREAD

**Discussion by bmcnellis** [JenniNiku/gllvm#147] [Q&A] [answered]
I'm trying to fit a beta-family gllvm to analyze some plant cover data. I get the following error:

```
Error in t.default(cbind(fit.mva$params$beta0, fit.mva$params$Xcoef)) : 
  argument is not a matrix
In addition: Warning messages:
1: In nlminb(objr$par, objr$fn, objr$gr, control = list(rel.tol = reltol,  :
  NA/NaN function evaluation
2: In gllvm.TMB(y = y, family = family, num.lv = 0, starting.val = "zero",  :
  Error in nlminb(objr$par, objr$fn, objr$gr, control = list(rel.tol = reltol,  : 
  NA/NaN gradient evaluation
```

I traced the error back to a matrix operation that is failing because the `params` entry within `fit.mva` doesn't exist - its not clear to me if this is a model specification issue or other code issue. Is there some other function input that is required for beta models? Or another vignette using them specifically? Thanks in advance!

Reproducible example:

```
devtools::install_github("admahood/neonPlantEcology")
data_div <- neonPlantEcology::npe_download(sites = "WREF")
data_df <- neonPlantEcology::npe_longform(data_div, scale = 'plot')
gllvm::gllvm(formula = cover ~ 1, data = data_df, family = "beta", num.lv = 2L, link = 'logit')
```

EDIT: the example data works if specified differently, and if adjusted to the 0-1 scale:

```
devtools::install_github("admahood/neonPlantEcology")
data_div <- neonPlantEcology::npe_download(sites = "WREF")
data_df <- neonPlantEcology::npe_longform(data_div, scale = 'plot')
data_mat <- neonPlantEcology::npe_community_matrix(data_div)
gllvm::gllvm(y = data_mat, family = "beta", num.lv = 2L, link = 'logit', starting.val = 'zero')
```

The error was quite cryptic before - perhaps something else would do better?

**Comment by BertvanderVeen** [comment, marked as answer]
Hello Brandon! Thanks for using the package, for posting your question, and for providing a reproducible example.

I have transferred your issue to the discussion forum, as there is no bug to report here.

The error you are experiencing is caused by the inability of the package to generate starting values. In part, this might be caused by a misunderstanding; the beta distribution (by definition) is only suitable for data in the 0-1 range. For data outside of that range, you can consider either a gamma or tweedie distribution for your responses instead.

On the other side, the package only allows specifying data in long format in a very particular manner (see e.g., https://cran.r-project.org/web/packages/gllvm/vignettes/vignette1.html);  as a dataframe instead of tibble, the first column for as numeric vector of species indices, and the way the data was provided in your first example does not fit that format.

Happy to help if you have any other questions or issues.

**Reply by bmcnellis** [reply]
Thanks very much for the answer Bert! I only just realized the issue was simple after exhausting all other much more complex solutions to the problem and assuming it was coded incorrectly. As is tradition, I suppose!

Perhaps an extra input check with an informative error message in the `gllvm` function would be helpful for others just starting out with the package? e.g.

```
if (family == "beta") {
      if (sum(apply(y, c(1, 2), \(xx) xx > 1 | xx < 0)) > 0) {
        stop("Input data for beta distribution must be in range 0 to 1")
      }
}
```

Or something, I'm sure there's a faster way to check

**Reply by BertvanderVeen** [reply]
Sure, we can add some defensive programming, thanks for the suggestion. See https://github.com/JenniNiku/gllvm/commit/03d3481bd6d7c37358c7183708e1a9489262ac43.
